lib/index.js.flow

```diff
@@ -0,0 +1,69 @@
+/* @flow */
+type ConnectAll = <D, P, S, C: React$Component<D, P, S>, SP, DP, Dispatch: Function>(
+  mapStateToProps: (state: Object) => SP,
+  mapDispatchToProps: (dispatch: Dispatch) => DP,
+  options?: {pure?: boolean, withRef?: boolean}
+) => (component: Class<C>) => Class<React$Component<D, $Diff<$Diff<P, DP>, SP>, S>>
+
+type ConnectAllStateless = <P, SP, DP, Dispatch: Function>(
+  mapStateToProps: (state: Object) => SP,
+  mapDispatchToProps: (dispatch: Dispatch) => DP,
+  options?: {pure?: boolean, withRef?: boolean}
+) => (component: (props: P) => any) => Class<React$Component<void, $Diff<$Diff<P, DP>, SP>, void>>
+
+type ConnectMerged = <D, P, S, C: React$Component<D, P, S>, SP, DP, MP, Dispatch: Function>(
+  mapStateToProps: (state: Object) => SP,
+  mapDispatchToProps: (dispatch: Dispatch) => DP,
+  mergeProps: (stateProps: SP, dispatchProps: DP, origProps: P) => MP,
+  options?: {pure?: boolean, withRef?: boolean}
+) => (component: Class<C>) => Class<React$Component<D, $Diff<P, MP>, S>>
+
+type ConnectMergedStateless = <P, SP, DP, MP, Dispatch: Function>(
+  mapStateToProps: (state: Object) => SP,
+  mapDispatchToProps: (dispatch: Dispatch) => DP,
+  mergeProps: (stateProps: SP, dispatchProps: DP, origProps: P) => MP,
+  options?: {pure?: boolean, withRef?: boolean}
+) => (component: (props: P) => any) => Class<React$Component<void, $Diff<P, MP>, void>>
+
+type ConnectNoState = <D, P, S, C: React$Component<D, P, S>, DP, Dispatch: Function>(
+    mapStateToProps: null | void,
+    mapDispatchToProps: (dispatch: Dispatch) => DP,
+    options?: {pure?: boolean, withRef?: boolean}
+  ) => (component: Class<C>) => Class<React$Component<D, $Diff<P, DP>, S>>
+
+type ConnectNoStateStatless = <P, DP, Dispatch: Function>(
+    mapStateToProps: null | void,
+    mapDispatchToProps: (dispatch: Dispatch) => DP,
+    options?: {pure?: boolean, withRef?: boolean}
+  ) => (component: (props: P) => any) => Class<React$Component<void, $Diff<P, DP>, void>>
+
+type ConnectDisptch = <D, P, S, C: React$Component<D, P, S>, SP, Dispatch: Function>(
+  mapStateToProps: (state: Object) => SP,
+  options?: {pure?: boolean, withRef?: boolean}
+) => (component: Class<C>) => Class<React$Component<D, $Diff<$Diff<P, {dispatch: Dispatch}>, SP>, S>>
+
+type ConnectDisptchStateless = <P, SP, Dispatch: Function>(
+  mapStateToProps: (state: Object) => SP,
+  options?: {pure?: boolean, withRef?: boolean}
```

Comment:
Oh I didn't know that. Still the overloads are still needed because every combination changes the type of the returned value. Every overload in the file has a different return type! I'll change the type defs though.